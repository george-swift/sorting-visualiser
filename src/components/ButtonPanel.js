import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';

const ButtonPanel = ({
  randomize, insertion, merge, quick, heap, sorted,
}) => {
  const [visible, setVisible] = useState(false);

  const startSorting = (algorithm) => {
    if (sorted) return;
    algorithm();
    setVisible(true);
  };

  useEffect(() => {
    if (sorted) setVisible(false);
  }, [sorted]);

  return (
    <div className="container button-panel">
      <button
        type="button"
        className="btn vsx-btn"
        onClick={randomize}
        disabled={visible}
      >
        Generate New List
      </button>
      <button
        type="button"
        className="btn vsx-btn"
        title="Worst case: O(N^2)"
        onClick={() => startSorting(insertion)}
        disabled={visible}
      >
        Insertion Sort
      </button>
      <button
        type="button"
        className="btn vsx-btn"
        title="Worst case: O(NLogN)"
        onClick={() => startSorting(merge)}
        disabled={visible}
      >
        Merge Sort
      </button>
      <button
        type="button"
        className="btn vsx-btn"
        title="Worst case: O(N^2)"
        onClick={() => startSorting(quick)}
        disabled={visible}
      >
        Quick Sort
      </button>
      <button
        type="button"
        className="btn vsx-btn"
        title="Worst case: O(NLogN)"
        onClick={() => startSorting(heap)}
        disabled={visible}
      >
        Heap Sort
      </button>
    </div>
  );
};

ButtonPanel.propTypes = {
  randomize: PropTypes.func.isRequired,
  insertion: PropTypes.func.isRequired,
  merge: PropTypes.func.isRequired,
  quick: PropTypes.func.isRequired,
  heap: PropTypes.func.isRequired,
  sorted: PropTypes.bool.isRequired,
};

export default ButtonPanel;
