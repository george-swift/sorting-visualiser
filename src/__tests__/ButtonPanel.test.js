import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import ButtonPanel from '../components/ButtonPanel';

const ButtonPanelProps = {
  algorithm: () => {},
  sorted: false,
};

describe('Snapshot of the ButtonPanel component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <ButtonPanel
        randomize={ButtonPanelProps.algorithm}
        insertion={ButtonPanelProps.algorithm}
        merge={ButtonPanelProps.algorithm}
        quick={ButtonPanelProps.algorithm}
        heap={ButtonPanelProps.algorithm}
        sorted={ButtonPanelProps.sorted}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('Funcitonality of buttons in panel', () => {
  beforeAll(() => {
    render(
      <ButtonPanel
        randomize={ButtonPanelProps.algorithm}
        insertion={ButtonPanelProps.algorithm}
        merge={ButtonPanelProps.algorithm}
        quick={ButtonPanelProps.algorithm}
        heap={ButtonPanelProps.algorithm}
        sorted={ButtonPanelProps.sorted}
      />,
    );
  });

  it('should change button state to disabled when sorting', () => {
    const mergeSortButton = screen
      .getAllByRole('button')
      .find((button) => button.textContent === 'Merge Sort');

    fireEvent.click(mergeSortButton);
    screen.getAllByRole('button').forEach((button) => {
      expect(button).toHaveAttribute('disabled');
    });
  });
});
