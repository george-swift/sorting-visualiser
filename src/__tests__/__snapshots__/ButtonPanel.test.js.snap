// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of the ButtonPanel component should render correctly 1`] = `
<div
  className="container button-panel"
>
  <button
    className="btn vsx-btn"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Generate New List
  </button>
  <button
    className="btn vsx-btn"
    disabled={false}
    onClick={[Function]}
    title="Worst case: O(N^2)"
    type="button"
  >
    Insertion Sort
  </button>
  <button
    className="btn vsx-btn"
    disabled={false}
    onClick={[Function]}
    title="Worst case: O(NLogN)"
    type="button"
  >
    Merge Sort
  </button>
  <button
    className="btn vsx-btn"
    disabled={false}
    onClick={[Function]}
    title="Worst case: O(N^2)"
    type="button"
  >
    Quick Sort
  </button>
  <button
    className="btn vsx-btn"
    disabled={false}
    onClick={[Function]}
    title="Worst case: O(NLogN)"
    type="button"
  >
    Heap Sort
  </button>
</div>
`;
